File: bindings/csharp/Serialization/BinaryWriter.cs
using System;

namespace MikanXR
{
	public class BinaryWriter
	{
		private static readonly int DefaultCapacity = 1024;

		private int _bytesWritten;
		private int _capacity;
		private byte[] _writeBuffer;

		public BinaryWriter()
		{
			_bytesWritten= 0;
			_capacity= DefaultCapacity;
			_writeBuffer= new byte[_capacity];
		}

		private void EnsureCapacity(int additionalBytes)
		{
			if (_bytesWritten + additionalBytes > _capacity)
			{
				_capacity= Math.Max(_capacity * 2, _bytesWritten + additionalBytes);
				Array.Resize(ref _writeBuffer, _capacity);
			}
		}

		public void Write(bool value)
		{
			Write(value ? (byte)1 : (byte)0);
		}

		public void Write(sbyte value)
		{
			EnsureCapacity(sizeof(sbyte));
			_writeBuffer[_bytesWritten]= unchecked((byte)value);
			_bytesWritten += sizeof(byte);
		}

		public void Write(byte value)
		{
			EnsureCapacity(sizeof(byte));
			_writeBuffer[_bytesWritten]= value;
			_bytesWritten += sizeof(byte);
		}

		public void Write(short value)
		{
			EnsureCapacity(sizeof(short));
			byte[] valueBytes = BitConverter.GetBytes(value);
			valueBytes.CopyTo(_writeBuffer, _bytesWritten);
			_bytesWritten += sizeof(short);
		}

		public void Write(ushort value)
		{
			EnsureCapacity(sizeof(ushort));
			byte[] valueBytes = BitConverter.GetBytes(value);
			valueBytes.CopyTo(_writeBuffer, _bytesWritten);
			_bytesWritten += sizeof(ushort);
		}

		public void Write(int value)
		{
			EnsureCapacity(sizeof(int));
			byte[] valueBytes = BitConverter.GetBytes(value);
			valueBytes.CopyTo(_writeBuffer, _bytesWritten);
			_bytesWritten += sizeof(int);
		}

		public void Write(uint value)
		{
			EnsureCapacity(sizeof(uint));
			byte[] valueBytes = BitConverter.GetBytes(value);
			valueBytes.CopyTo(_writeBuffer, _bytesWritten);
			_bytesWritten += sizeof(uint);
		}

		public void Write(ulong value)
		{
			EnsureCapacity(sizeof(ulong));
			byte[] valueBytes = BitConverter.GetBytes(value);
			valueBytes.CopyTo(_writeBuffer, _bytesWritten);
			_bytesWritten += sizeof(ulong);
		}

		public void Write(float value)
		{
			EnsureCapacity(sizeof(float));
			byte[] floatBytes = BitConverter.GetBytes(value);
			floatBytes.CopyTo(_writeBuffer, _bytesWritten);
			_bytesWritten += sizeof(float);
		}

		public void Write(double value)
		{
			EnsureCapacity(sizeof(double));
			byte[] doubleBytes = BitConverter.GetBytes(value);
			doubleBytes.CopyTo(_writeBuffer, _bytesWritten);
			_bytesWritten += sizeof(double);
		}

		public void WriteUTF8String(string unicodeString)
		{
			byte[] utf8Bytes = System.Text.Encoding.UTF8.GetBytes(unicodeString);

			Int32 stringLength = utf8Bytes.Length;
			Write(stringLength);

			if (stringLength > 0)
			{
				Write(utf8Bytes);
			}
		}

		public void Write(byte[] value)
		{
			EnsureCapacity(value.Length);
			value.CopyTo(_writeBuffer, _bytesWritten);
			_bytesWritten += value.Length;
		}

		public byte[] ToArray()
		{
			if (_writeBuffer.Length != _bytesWritten)
			{
				Array.Resize(ref _writeBuffer, _bytesWritten);
			}

			return _writeBuffer;
		}
	}
}

File: bindings/typescript/Serialization/BinaryWriter.ts
/**
 * Binary writer for serializing data to binary format
 * Writes in little-endian format to match C# implementation
 */
export class BinaryWriter {
  private static readonly DEFAULT_CAPACITY = 1024;

  private bytesWritten: number = 0;
  private capacity: number;
  private writeBuffer: Uint8Array;
  private dataView: DataView;

  constructor() {
    this.capacity = BinaryWriter.DEFAULT_CAPACITY;
    this.writeBuffer = new Uint8Array(this.capacity);
    this.dataView = new DataView(this.writeBuffer.buffer);
  }

  private ensureCapacity(additionalBytes: number): void {
    if (this.bytesWritten + additionalBytes > this.capacity) {
      this.capacity = Math.max(this.capacity * 2, this.bytesWritten + additionalBytes);
      const newBuffer = new Uint8Array(this.capacity);
      newBuffer.set(this.writeBuffer);
      this.writeBuffer = newBuffer;
      this.dataView = new DataView(this.writeBuffer.buffer);
    }
  }

  write(value: boolean): void;
  write(value: number): void;
  write(value: bigint): void;
  write(value: Uint8Array): void;
  write(value: boolean | number | bigint | Uint8Array): void {
    if (typeof value === 'boolean') {
      this.writeBoolean(value);
    } else if (typeof value === 'number') {
      this.writeByte(value);
    } else if (typeof value === 'bigint') {
      this.writeInt64(value);
    } else if (value instanceof Uint8Array) {
      this.writeBytes(value);
    }
  }

  writeBoolean(value: boolean): void {
    this.writeByte(value ? 1 : 0);
  }

  writeSByte(value: number): void {
    this.ensureCapacity(1);
    this.dataView.setInt8(this.bytesWritten, value);
    this.bytesWritten += 1;
  }

  writeByte(value: number): void {
    this.ensureCapacity(1);
    this.dataView.setUint8(this.bytesWritten, value);
    this.bytesWritten += 1;
  }

  writeInt16(value: number): void {
    this.ensureCapacity(2);
    this.dataView.setInt16(this.bytesWritten, value, true); // true = little endian
    this.bytesWritten += 2;
  }

  writeUInt16(value: number): void {
    this.ensureCapacity(2);
    this.dataView.setUint16(this.bytesWritten, value, true);
    this.bytesWritten += 2;
  }

  writeInt32(value: number): void {
    this.ensureCapacity(4);
    this.dataView.setInt32(this.bytesWritten, value, true);
    this.bytesWritten += 4;
  }

  writeUInt32(value: number): void {
    this.ensureCapacity(4);
    this.dataView.setUint32(this.bytesWritten, value, true);
    this.bytesWritten += 4;
  }

  writeInt64(value: bigint): void {
    this.ensureCapacity(8);
    this.dataView.setBigInt64(this.bytesWritten, value, true);
    this.bytesWritten += 8;
  }

  writeUInt64(value: bigint): void {
    this.ensureCapacity(8);
    this.dataView.setBigUint64(this.bytesWritten, value, true);
    this.bytesWritten += 8;
  }

  writeFloat(value: number): void {
    this.ensureCapacity(4);
    this.dataView.setFloat32(this.bytesWritten, value, true);
    this.bytesWritten += 4;
  }

  writeDouble(value: number): void {
    this.ensureCapacity(8);
    this.dataView.setFloat64(this.bytesWritten, value, true);
    this.bytesWritten += 8;
  }

  writeUTF8String(value: string): void {
    const encoder = new TextEncoder();
    const utf8Bytes = encoder.encode(value);

    const stringLength = utf8Bytes.length;
    this.writeInt32(stringLength);

    if (stringLength > 0) {
      this.writeBytes(utf8Bytes);
    }
  }

  writeBytes(value: Uint8Array): void {
    this.ensureCapacity(value.length);
    this.writeBuffer.set(value, this.bytesWritten);
    this.bytesWritten += value.length;
  }

  toArray(): Uint8Array {
    return this.writeBuffer.slice(0, this.bytesWritten);
  }
}
